test/addons-napi/6_object_wrap/myobject.cc

```diff
@@ -0,0 +1,199 @@
+#include "myobject.h"
+
+napi_ref MyObject::constructor;
+
+MyObject::MyObject(double value)
+    : value_(value), env_(nullptr), wrapper_(nullptr) {}
+
+MyObject::~MyObject() { napi_delete_reference(env_, wrapper_); }
+
+void MyObject::Destructor(void* nativeObject, void* /*finalize_hint*/) {
+  reinterpret_cast<MyObject*>(nativeObject)->~MyObject();
+}
+
+#define DECLARE_NAPI_METHOD(name, func)                          \
+  { name, func, 0, 0, 0, napi_default, 0 }
+
+void MyObject::Init(napi_env env, napi_value exports) {
+  napi_status status;
+  napi_property_descriptor properties[] = {
+      { "value", nullptr, GetValue, SetValue, 0, napi_default, 0 },
+      DECLARE_NAPI_METHOD("plusOne", PlusOne),
+      DECLARE_NAPI_METHOD("multiply", Multiply),
+  };
+
+  napi_value cons;
+  status =
+      napi_define_class(env, "MyObject", New, nullptr, 3, properties, &cons);
+  if (status != napi_ok) return;
+
+  status = napi_create_reference(env, cons, 1, &constructor);
+  if (status != napi_ok) return;
+
+  status = napi_set_named_property(env, exports, "MyObject", cons);
+  if (status != napi_ok) return;
+}
+
+void MyObject::New(napi_env env, napi_callback_info info) {
+  napi_status status;
+
+  bool is_constructor;
+  status = napi_is_construct_call(env, info, &is_constructor);
+  if (status != napi_ok) return;
+
+  if (is_constructor) {
+    // Invoked as constructor: `new MyObject(...)`
+    napi_value args[1];
+    status = napi_get_cb_args(env, info, args, 1);
+    if (status != napi_ok) return;
+
+    double value = 0;
+
+    napi_valuetype valuetype;
+    status = napi_typeof(env, args[0], &valuetype);
+    if (status != napi_ok) return;
+
+    if (valuetype != napi_undefined) {
+      status = napi_get_value_double(env, args[0], &value);
+      if (status != napi_ok) return;
+    }
+
+    MyObject* obj = new MyObject(value);
+
+    napi_value jsthis;
+    status = napi_get_cb_this(env, info, &jsthis);
+    if (status != napi_ok) return;
+
+    obj->env_ = env;
+    status = napi_wrap(env,
+                       jsthis,
+                       reinterpret_cast<void*>(obj),
```

Comment:
You shouldn’t need a `reinterpret_cast` here, and iirc not even any cast at all?